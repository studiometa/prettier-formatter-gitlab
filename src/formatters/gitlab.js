import { createHash } from 'node:crypto';
import { showInvisibles, generateDifferences } from 'prettier-linter-helpers';
import { getPrettierFileInfos } from '../utils/get-prettier-file-infos.js';

/**
 * @typedef {import('../types.d.ts').FileInfo} FileInfo
 * @typedef {import('../types.d.ts').CodeQualityReport} CodeQualityReport
 */

/**
 * Create a fingerprint for the given file and error message.
 * @param  {string[]} args
 * @returns {string}
 */
function createFingerprint(...args) {
  const md5 = createHash('md5');
  args.forEach((arg) => {
    md5.update(String(arg));
  });
  return md5.digest('hex');
}

/**
 * Format a file.
 * @param {Partial<FileInfo>} fileInfo
 * @returns {CodeQualityReport[]}
 */
function formatFile({ filename, input, output, error }) {
  const { INSERT, DELETE, REPLACE } = generateDifferences;

  if (error) {
    return [
      {
        type: 'issue',
        check_name: 'prettier',
        description: String(error),
        severity: 'major',
        fingerprint: createFingerprint(filename, error.message),
        location: {
          path: filename,
        },
      },
    ];
  }

  const differences = generateDifferences(input, output);

  return differences.map(({ offset, operation, deleteText = '', insertText = '' }) => {
    const deleteCode = showInvisibles(deleteText);
    const insertCode = showInvisibles(insertText);
    const begin = input.slice(0, offset).split('\n').length;
    const endOffset = deleteText.split('\n').length - 1;
    const end = begin + endOffset;

    let message;

    switch (operation) {
      case INSERT:
        message = `Insert ${insertCode}`;
        break;
      case DELETE:
        message = `Delete ${deleteCode}`;
        break;
      case REPLACE:
        message = `Replace ${deleteCode} with ${insertCode}`;
        break;
    }

    return {
      type: 'issue',

      check_name: 'prettier',
      description: message,
      severity: 'minor',
      fingerprint: createFingerprint(filename, message, begin, end),
      location: {
        path: filename,
        lines: {
          begin,
          end,
        },
      },
    };
  });
}

/**
 * GitLab formatter.
 * @param   {string[]} files
 * @returns {Promise<CodeQualityReport[]>}
 */
export async function gitlab(files) {
  const infos = await Promise.all(files.map((file) => getPrettierFileInfos(file)));
  return infos.reduce((acc, fileInfo) => [...acc, ...formatFile(fileInfo)], []);
}
