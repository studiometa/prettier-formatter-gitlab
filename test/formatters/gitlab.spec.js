import { describe, it, expect } from 'bun:test';
import { resolve } from 'node:path';
import { gitlab } from '../../src/formatters/gitlab.js';

describe('The gitlab formatter', () => {
  it('should return gitlab formatted reports about the given files', async () => {
    const filepath = resolve(import.meta.dirname, '../__stubs__/dirty.js');
    const results = await gitlab([filepath]);
    expect(results).toMatchSnapshot();
  });

  it('should handle files with syntax errors', async () => {
    const filepath = resolve(import.meta.dirname, '../__stubs__/syntax-error.js');
    const results = await gitlab([filepath]);
    expect(results).toMatchSnapshot();
  });
});
