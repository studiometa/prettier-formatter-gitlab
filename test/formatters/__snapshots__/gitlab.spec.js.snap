// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`The gitlab formatter should return gitlab formatted reports about the given files 1`] = `
[
  {
    "check_name": "prettier",
    "description": "Replace ⏎⏎⏎⏎⏎⏎function·foo(·arg··)·{⏎⏎ with function·foo(arg)·{",
    "fingerprint": "c38522eb22f56c446d6b78774d940d24",
    "location": {
      "lines": {
        "begin": 3,
        "end": 11,
      },
      "path": "test/__stubs__/dirty.js",
    },
    "severity": "minor",
    "type": "issue",
  },
  {
    "check_name": "prettier",
    "description": "Delete ⏎⏎⏎",
    "fingerprint": "f312c866f5c2332c733d2de24b9efa0f",
    "location": {
      "lines": {
        "begin": 12,
        "end": 15,
      },
      "path": "test/__stubs__/dirty.js",
    },
    "severity": "minor",
    "type": "issue",
  },
]
`;

exports[`The gitlab formatter should handle files with syntax errors 1`] = `
[
  {
    "check_name": "prettier",
    "description": 
"SyntaxError: Unexpected token, expected "(" (3:13)
\x1B[0m \x1B[90m 1 |\x1B[39m \x1B[90m/* eslint-disable */\x1B[39m
 \x1B[90m 2 |\x1B[39m
\x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 3 |\x1B[39m \x1B[36mfunction\x1B[39m foo\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m( arg  ) {
 \x1B[90m   |\x1B[39m             \x1B[31m\x1B[1m^\x1B[22m\x1B[39m
 \x1B[90m 4 |\x1B[39m   \x1B[36mreturn\x1B[39m arg\x1B[33m;\x1B[39m
 \x1B[90m 5 |\x1B[39m }
 \x1B[90m 6 |\x1B[39m\x1B[0m"
,
    "fingerprint": "462882d90971b8c9dcaaa2b4d74d1da6",
    "location": {
      "path": "test/__stubs__/syntax-error.js",
    },
    "severity": "major",
    "type": "issue",
  },
]
`;
